sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/db/migration/V8520__Attachments.java

```diff
@@ -0,0 +1,177 @@
+      throw new IllegalStateException("MISO_FILES_DIR not set");
+    }
+    if (!filesDir.endsWith(File.separator)) {
+      filesDir += File.separator;
+    }
+    basePath = Paths.get(filesDir);
+
+    log.info("scanning directory {}", filesDir);
+    try (PreparedStatement insertStatement = connection.prepareStatement(
+        "INSERT INTO Attachment(filename, path, creator, created)"
+            + " VALUES (?, ?, (SELECT userId FROM User WHERE loginName = 'admin'), NOW());",
+        Statement.RETURN_GENERATED_KEYS)) {
+      for (EntityType type : EntityType.values()) {
+        scanTypeDir(type, insertStatement);
+      }
+    }
+
+    dropTempTable();
+  }
+
+  private void scanTypeDir(EntityType type, PreparedStatement insertStatement) throws SQLException {
+    File dir = new File(filesDir + type.getDirName());
+    if (dir.exists()) {
+      log.info("scanning directory {}", dir.getAbsolutePath());
+      try (PreparedStatement getStatement = connection
+          .prepareStatement(String.format("SELECT * FROM %s WHERE %s = ?", type.getTable(), type.getPrimaryKey()));
+          PreparedStatement joinStatement = connection.prepareStatement(
+              String.format("INSERT INTO %s(%s, attachmentId) VALUES (?, ?);", type.getJoinTable(), type.getPrimaryKey()))) {
+        for (File typeFile : dir.listFiles()) {
+          processObjectDir(type, typeFile, getStatement, insertStatement, joinStatement);
+        }
+      }
+    }
+  }
+
+  private void processObjectDir(EntityType type, File dir, PreparedStatement getStatement, PreparedStatement insertStatement,
+      PreparedStatement joinStatement) throws SQLException {
+    log.info("scanning directory " + dir.getAbsolutePath());
+    if (!dir.isDirectory()) {
+      log.warn("Unexpected non-directory found: {}", dir.getAbsolutePath());
+    } else {
+      long entityId = 0L;
+      try {
+        entityId = Long.parseLong(dir.getName());
+      } catch (NumberFormatException e) {
+        // ignore error - failure handled below
+      }
+      if (entityId == 0L) {
+        log.warn("Unexpected directory name: {}", dir.getAbsolutePath());
+      } else {
+        getStatement.setLong(1, entityId);
+        try (ResultSet results = getStatement.executeQuery()) {
+          if (!results.next()) {
+            log.warn("Found files for non-existant {} {}", type.getTable(), entityId);
+          } else {
+            joinStatement.setLong(1, entityId);
+            for (File objectFile : dir.listFiles()) {
+              processFile(objectFile, insertStatement, joinStatement);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private void processFile(File objectFile, PreparedStatement insertStatement, PreparedStatement joinStatement) throws SQLException {
+    if (!objectFile.isFile()) {
+      log.warn("Unexpected non-file found: {}", objectFile.getAbsolutePath());
+    } else {
+      log.info("processing file " + objectFile.getAbsolutePath());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Format specifiers should be used instead of string concatenation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)